### PR TITLE
add container runtime for pipeline

### LINKED ISSUE #1334: Implement support for docker function execution
part of #1279 

The docker runtime implementation will be put in kpt repo instead of kyaml

### PR DESCRIPTION
**Pull request by Shell32-Natsu**
close #1334

### CODE REVIEW

File: internal/pipeline/runtime/container.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	AsCurrentUser bool
+}
+
+// ContainerRunner implements a KRMFn which run a containerized
+// KRM function
+type ContainerRunner struct {
+	// Image is the container image to run
+	Image string
+
+	Exec ExecRunner
```
Comment (droot) [inline internal/pipeline/runtime/container.go:42]: Not sure why we have this here ?

File: internal/pipeline/runtime/container.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+// ContainerNetworkName is a type for network name used in container
+type containerNetworkName string
+
+const (
+	networkNameNone containerNetworkName = "none"
+	networkNameHost containerNetworkName = "host"
+)
+
+type containerFunctionPermission struct {
```
Comment (droot) [inline internal/pipeline/runtime/container.go:31]: May be shorten it `containerFnPermission` ?

File: internal/pipeline/runtime/container.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	AllowNetwork  bool
+	AsCurrentUser bool
+}
+
+// ContainerRunner implements a KRMFn which run a containerized
+// KRM function
+type ContainerRunner struct {
```
Comment (droot) [inline internal/pipeline/runtime/container.go:38]: Thinking of naming here: would calling it `ContainerFn` be more appropriate ?

File: internal/pipeline/runtime/container.go (hunk 1 above)
Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container.go:42]: 1. This piece of code is modified from kyaml function runtime and the container runtime in kyaml is based on exec runtime.
2. Since we have plan to support exec function so I think it's better to have a separate runner for exec so I don't change this structure.

File: internal/pipeline/runtime/container.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	// Image is the container image to run
+	Image string
+
+	Exec ExecRunner
+
+	containerFunctionPermission
```
Comment (droot) [inline internal/pipeline/runtime/container.go:44]: Any reasons for preferring embedding vs explicit field `perms` ? I do want to refresh my notes on this topic.

File: internal/pipeline/runtime/exec.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	Args []string `yaml:"args,omitempty"`
+}
+
+// Run runs the command
+func (c *ExecRunner) Run(reader io.Reader, writer io.Writer) error {
+	cmd := exec.Command(c.Path, c.Args...)
+	cmd.Stdin = reader
+	cmd.Stdout = writer
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
```
Comment (droot) [inline internal/pipeline/runtime/exec.go:38]: We should use `ctx` that allows to set timeout etc. For now pick a large timeout but we should do it now avoid it plumbing it later.

File: internal/pipeline/runtime/container.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		return fmt.Errorf("error when get command and args: %w", err)
+	}
+	f.Exec.Path = path
+	f.Exec.Args = args
+	return nil
+}
+
+// getCommand returns the command + args to run to spawn the container
+func (f *ContainerRunner) getCommand() (string, []string, error) {
```
Comment (droot) [inline internal/pipeline/runtime/container.go:73]: may be have this function return `exec.Cmd`

**Review by droot — CHANGES_REQUESTED** [review]
Took a quick look. Some early comments.

File: internal/pipeline/runtime/container.go (hunk 6 above)
Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container.go:73]: I think return an ExecFn may be better? Or just merge this function with `setupExec` and directly setup the `Exec` member. `exec.Cmd` should be encapsulated by `ExecFn`.

File: internal/pipeline/runtime/container.go (hunk 1 above)
Comment (droot) [reply internal/pipeline/runtime/container.go]: My main motivation for that suggestion was decoupling the execFn and containerFn so that changes to one doesn't affect the another. Also note that the execFn is a very lightweight wrapper over golang's exec primitives, so this extra layering doesn't buy us much instead increases overhead for anyone understanding `containerFn`.

I know current code is inspired by the code in kyaml but we can diverge from that depending on our requirements.

> Since we have plan to support exec function so I think it's better to have a separate runner for exec so I don't change this structure.

Totally agree with having separate runner for execFn. I am suggesting even greater degree of separation between the two :)

File: internal/pipeline/runtime/container.go (hunk 6 above)
Comment (droot) [reply internal/pipeline/runtime/container.go]: I shared my view in an earlier comment. If we decide to go that route, this can be revisited.

File: internal/pipeline/runtime/exec.go (hunk 5 above)
Comment (droot) [reply internal/pipeline/runtime/exec.go:53]: I realize, we will have to change `KRMFn` to accept `ctx` and relay that here.

File: internal/pipeline/runtime/exec_test.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/internal/pipeline/runtime"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestExecRunner(t *testing.T) {
```
Comment (droot) [inline internal/pipeline/runtime/exec_test.go:25]: We can add test for `containerFn` that requires docker on the machine and we can use Golang's build tags to selectively invoke that. You can read more about https://mickey.dev/posts/go-build-tags-testing/

**Review by droot — CHANGES_REQUESTED** [review]

File: internal/pipeline/runtime/container.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	Image string
+	// Container function will be killed after this timeour.
+	// The default value is 5 minutes.
+	Timeout time.Duration
+	Perm    ContainerFnPermission
+}
+
+// Run implements KRMFn. It will run the container function with
+// stdin from r and write the output to w
+func (f *ContainerFn) Run(r io.Reader, w io.Writer) error {
```
Comment (frankfarzan) [inline internal/pipeline/runtime/container.go:60]: Are you implementing this to work on Windows?

File: internal/pipeline/runtime/container_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"bytes"
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/internal/pipeline/runtime"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestContainerFn(t *testing.T) {
+	// use sed to test instead of calling real docker
```
Comment (frankfarzan) [inline internal/pipeline/runtime/container_test.go:28]: how do you use `sed`?

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: internal/pipeline/runtime/container_test.go (hunk 9 above)
Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container_test.go:28]: Updated the test but forgot this comment which was copied from kyaml. Will remove it.

File: internal/pipeline/runtime/container.go (hunk 8 above)
Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container.go:60]: @droot Do we want to support Windows?

Comment (droot) [reply internal/pipeline/runtime/container.go:60]: That's a good question. Didn't think about this. We would like to support windows, but reality is testing it regularly requires access to windows machines etc and I have found that to be challenging. Actual work required to support it won't be much IMO. From my previous project experience, what has helped is to create an issue for windows support and our friends from Windows eco-system help to support that.

TL;DR: windows support will be best effort basis and not a priority at this point.

Comment (frankfarzan) [reply internal/pipeline/runtime/container.go:60]: I think we should avoid regression if kpt supports windows today, but it can be punted for now and tracked as a support issue.

Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container.go:60]: https://github.com/GoogleContainerTools/kpt/issues/1367

File: internal/pipeline/runtime/container_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
+// +build docker
```
[the remark this answers is not in the document]
Comment (Shell32-Natsu) [reply internal/pipeline/runtime/container_test.go:1]: done

**Comment by Shell32-Natsu** [comment]
@droot @frankfarzan PTAL

**Review by droot — APPROVED** [review]

**Review by frankfarzan — APPROVED** [review]
